File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/tournament/service/TournamentService.java
package com.cwtsite.cwt.domain.tournament.service;

import com.cwtsite.cwt.domain.application.service.ApplicationRepository;
import com.cwtsite.cwt.domain.core.exception.ResourceNotFoundException;
import com.cwtsite.cwt.domain.tournament.entity.Tournament;
import com.cwtsite.cwt.domain.tournament.entity.enumeration.TournamentStatus;
import com.cwtsite.cwt.domain.user.repository.UserRepository;
import com.cwtsite.cwt.domain.user.repository.entity.User;
import com.cwtsite.cwt.entity.Application;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Component
public class TournamentService {

    private final UserRepository userRepository;
    private final TournamentRepository tournamentRepository;
    private final ApplicationRepository applicationRepository;

    @Autowired
    public TournamentService(UserRepository userRepository, TournamentRepository tournamentRepository,
                             ApplicationRepository applicationRepository) {
        this.userRepository = userRepository;
        this.tournamentRepository = tournamentRepository;
        this.applicationRepository = applicationRepository;
    }

    public Tournament startNewTournament(final User host, final List<Long> moderatorIds) throws IllegalStateException {
        long numberOfUnarchivedTournaments = tournamentRepository.countByStatusNot(TournamentStatus.ARCHIVED);

        if (numberOfUnarchivedTournaments > 0) {
            throw new IllegalStateException("For a new tournament to start it is required that all tournaments are archived.");
        }

        final Set<User> moderators = moderatorIds == null || moderatorIds.isEmpty()
                ? Collections.emptySet()
                : new HashSet<>(userRepository.findAll(moderatorIds));

        Tournament tournament = new Tournament();

        tournament.getModerators().addAll(moderators);
        tournament.setHost(host);
        tournament.setStatus(TournamentStatus.OPEN);

        return tournamentRepository.save(tournament);
    }

    public Tournament getTournament(final long id) {
        return tournamentRepository.findOne(id);
    }

    public Tournament getCurrentTournament() {
        final Tournament currentTorunament = tournamentRepository.findByStatusNot(TournamentStatus.ARCHIVED);

        if (currentTorunament == null) {
            throw new ResourceNotFoundException("There is currently no tournament");
        }

        return currentTorunament;
    }

    public List<Application> getApplicants(Tournament tournament) {
        return this.applicationRepository.findByTournament(tournament);
    }

    public List<Tournament> getAll() {
        return tournamentRepository.findAll();
    }

    public Tournament getTournamentByYear(long year) {
        return tournamentRepository.findByYear((int) year);
    }
}

File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/tournament/service/TournamentService.kt
package com.cwtsite.cwt.domain.tournament.service

import com.cwtsite.cwt.domain.application.service.ApplicationRepository
import com.cwtsite.cwt.domain.game.entity.Game
import com.cwtsite.cwt.domain.game.service.GameRepository
import com.cwtsite.cwt.domain.tournament.entity.Tournament
import com.cwtsite.cwt.domain.tournament.entity.enumeration.TournamentStatus
import com.cwtsite.cwt.domain.user.repository.UserRepository
import com.cwtsite.cwt.entity.Application
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Component
class TournamentService @Autowired
constructor(private val userRepository: UserRepository, private val tournamentRepository: TournamentRepository,
            private val applicationRepository: ApplicationRepository, private val gameRepository: GameRepository) {

    @Throws(IllegalStateException::class)
    fun startNewTournament(moderatorIds: List<Long>): Tournament {
        val numberOfUnarchivedTournaments = tournamentRepository.countByStatusNot(TournamentStatus.ARCHIVED)

        if (numberOfUnarchivedTournaments > 0) {
            throw IllegalStateException("For a new tournament to start it is required that all tournaments are archived.")
        }

        val tournament = Tournament()
        tournament.moderators = userRepository.findAllById(moderatorIds).toSet()
        tournament.status = TournamentStatus.OPEN

        return tournamentRepository.save(tournament)
    }

    @Transactional
    fun startPlayoffs(games: List<Game>): List<Game> {
        val currentTournament = getCurrentTournament()
        currentTournament.status = TournamentStatus.PLAYOFFS
        tournamentRepository.save(currentTournament)
        return gameRepository.saveAll(games)
    }

    fun getTournament(id: Long): Optional<Tournament> {
        return tournamentRepository.findById(id)
    }

    fun getApplicants(tournament: Tournament): List<Application> {
        return this.applicationRepository.findByTournament(tournament)
    }

    fun getTournamentByYear(year: Long): Optional<Tournament> {
        return tournamentRepository.findByYear(year.toInt())
    }

    fun getCurrentTournament(): Tournament = tournamentRepository.findByStatusNot(TournamentStatus.ARCHIVED)
            ?: throw RuntimeException("There is currently no tournament.")

    fun getAll(): List<Tournament> = tournamentRepository.findAll()
}